### PR TITLE
Cleanup of grunt tasks, tests integrated. (rev1)

### PR DESCRIPTION
**Pull request by nodesocket**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -10,14 +10,14 @@
 
     git clone git@github.com:balanced/balanced-js.git
     cd balanced-js
-    npm install -g grunt-cli
-    npm install
+    sudo npm install -g grunt-cli
+    sudo npm install
```
Comment (mjallday) [inline README.md]: i don't sudo my install

File: README.md [hunk 2]
```diff
@@ -10,14 +10,14 @@
...
 
 ## Building ###
 
 ##### Both JS and Proxy #####
 
-`grunt` or `grunt build`
+    grunt or grunt build
```
Comment (mjallday) [inline README.md]: error?

File: README.md (hunk 1 above)
Comment (nodesocket) [inline README.md]: I had to `sudo` for istanbul

File: README.md (hunk 2 above)
Comment (nodesocket) [inline README.md]: Cleanup, standardize like the rest in < pre > tags.

File: README.md (hunk 1 above)
Comment (mjallday) [inline README.md]: maybe check your environment then, i didn't have to do it.

File: README.md (hunk 2 above)
Comment (mjallday) [inline README.md]: generally everything in the pre tags is a command you can run. e.g. `grunt` or `grunt build` are two distinct commands.

File: src/balanced.js [hunk 3]
```diff
@@ -1,3 +1,12 @@
+////
+// Required for ie < 9 support
+////
+if (!String.prototype.trim) {
+    String.prototype.trim=function() {
```
Comment (mjallday) [inline src/balanced.js:5]: we should have a test for this function

File: README.md (hunk 1 above)
Comment (nodesocket) [inline README.md]: Found the problem, needs to do:

```
sudo chown -R justin: ~/.npm
```

File: src/balanced.js (hunk 3 above)
Comment (nodesocket) [inline src/balanced.js:5]: Comes from http://stackoverflow.com/a/498995/425964

Comment (mjallday) [inline src/balanced.js:5]: great, just take the tests from there then.

File: test/unit/cards.js [hunk 4]
```diff
@@ -6,6 +6,471 @@ module('balanced.js.cards', {
...
+            expected: false
+        },
+    ];
+
+    for(var i = 0; i < tests.length; i++) {
+        assert.equal(balanced.card.isSecurityCodeValid(tests[i].number, tests[i].csc), tests[i].expected, "Test #" + (i + 1));
+    }
+});
+
+test('isExpiryValid', function (assert) {
```
Comment (mjallday) [inline test/unit/cards.js:284]: the method is expiration not expiry isn't it?

File: test/unit/emailAddress.js [hunk 5]
```diff
@@ -0,0 +1,72 @@
+module('balanced.js.emailAddress', {
+    setup: function () {
+    },
+
+    teardown: function () {
+    }
+});
+
+test('validate', function (assert) {
+    var tests = [
```
Comment (mjallday) [inline test/unit/emailAddress.js:10]: these would be way more compact if you wrote

```
var validEmails = [...];
var invalidEmails = [...]

for email in validEmails
   assert.true(balanced.emailAddress.validate(email);

for email in invalidEmails
   assert.false(balanced.emailAddress.validate(email);
```

not a biggie but something stylistic to think about

File: test/unit/cards.js (hunk 4 above)
Comment (nodesocket) [inline test/unit/cards.js:284]: in `src/balanced.js` it is:

```
isExpiryValid:function (expiryMonth, expiryYear) {
```

File: test/unit/emailAddress.js (hunk 5 above)
Comment (nodesocket) [inline test/unit/emailAddress.js:10]: I actually prefer the style I have, as its standard across all tests, using an array of objects with properties, and then an `expected` property.

**Comment by kleinsch** [comment]
Tests all pass, although the coverage threshold fails. I took a look at the coverage report and we're missing tests for:
- create (card)
- create (bankAccount)
- lookupRoutingNumber

Other than that, coverage report looks good. We're always going to have low coverage numbers because of the library code that gets included into balanced.js, so it would probably make sense to either tweak the coverage numbers down or change how we run the tests so coverage is just checking our code, not the libraries.

Ran the examples and it looks like it's trying to POST to localhost when creating anything, so that's all failing.

screen shot 2013-08-30 at 1 53 46 pm

`README.md` looks solid, I was able to follow all the instructions. One thing I realized on the dashboard project: if you use the `grunt-open` plugin, you can automatically open a browser window to the right address as part of `grunt serve`, which removes one step from the instructions.

**Comment by kleinsch** [comment]
One more thing... it would probably make sense for `grunt serve` to build whatever it needs. Right now if you check out from scratch and run `grunt serve` it spits out an error because nothing has built yet.

**Comment by nodesocket** [comment]
@kleinsch Thanks for the feedback. Do you know how to adjust Qunit to check only our code, not the libraries to increase the coverage?

I'll integrate `grunt-open` and update.

I went back and forth if `grunt serve` should **build**. Ultimately decided its best to leave those as two separate disjointed actions. I can change though, do you think it makes sense to call build in serve.

**Comment by kleinsch** [comment]
The qunit plugin runs coverage data on whatever is specified in the `src` config option for the `coverage` section. Since you're running the tests against the concatenated file, it's checking the whole thing. If you want to change what it uses for coverage, you'd have to change the tests to include each file individually, then only run coverage on the balanced portion. I think the way it's working right now is probably fine. It makes more sense to me to tweak the thresholds than to run the tests against something that's not the final product.

Yeah, I could go either way on `grunt serve` also, but I think it'd make it easier for first time users if it builds also. One less thing for people to forget.

**Pull request merged**